### PR TITLE
Add documentation

### LINKED ISSUE #34: TODO List Release
| Q | A |
| --- | --- |
| Bug? | no |
| New Feature? | yes |
- [x] Doc Full Configuration Example
- [x] Doc Create Login
- [x] Doc Create Registration
- [x] Doc Create Password Reset / Forget
- [x] Doc Create Profile
- [x] Doc Completion Listener
- [x] Doc Blacklisting
- [ ] #10

### PR DESCRIPTION
**Pull request by alexander-schranz**
| Q | A |
| --- | --- |
| Bug fix? | no |
| New feature? | yes |
| BC breaks? | no |
| Deprecations? | no |
| Fixed tickets | fixes #34 |
| License | MIT |
#### What's in this PR?

https://github.com/sulu/SuluCommunityBundle/tree/feature/doc

### CODE REVIEW

**Comment by wachterjohannes** [comment]
@danrot could you read the docs for typos? technically it is correct.

File: Resources/doc/10-email-confirmation.md [hunk 1]
```diff
@@ -0,0 +1,63 @@
+# Email Confirmation
+
+## Config
+
+
```
Comment (wachterjohannes) [inline Resources/doc/10-email-confirmation.md]: remove one empty line

File: Resources/doc/10-email-confirmation.md [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+{% endblock %}
+```
+
+## template
+
+Template which is render after successfully confirmed the email address.
+
+**Example Template**:
+
+
```
Comment (wachterjohannes) [inline Resources/doc/10-email-confirmation.md]: only one empty line

File: README.md [hunk 3]
```diff
@@ -1,230 +1,26 @@
...
-                path: sulu_community.logout
-                target: sulu_community.login
-            remember_me:
-                secret:   '%secret%'
-                lifetime: 604800 # 1 week in seconds
-                path:     /
+ - Login
+ - Registration
+ - Password Forget
+ - Password Reset
```
Comment (danrot) [inline README.md]: What's the difference between reset and forget?

File: Resources/doc/10-email-confirmation.md [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+{% block content %}
+    {% set confirmUrl = url('sulu_community.email_confirmation', { token: token }) %}
+    <a href="{{ confirmUrl }}">{{ confirmUrl }}</a>
+{% endblock %}
+```
+
+## template
+
+Template which is render after successfully confirmed the email address.
```
Comment (danrot) [inline Resources/doc/10-email-confirmation.md]: `which is rendered`

Comment (danrot) [inline Resources/doc/10-email-confirmation.md]: `after the email address has been successfully confirmed`

File: Resources/doc/11-completion.md [hunk 5]
```diff
@@ -0,0 +1,210 @@
+# Completion
+
+You can add an additional completion form to complete the user registration after confirmation.
```
Comment (danrot) [inline Resources/doc/11-completion.md]: `an additional form to complete the user registration` - having complete and completion is a bit redundant.

File: Resources/doc/11-completion.md [hunk 6]
```diff
@@ -0,0 +1,210 @@
...
+                    admin_template: ~
+                redirect_to: /
+                service: app.completion_validator
+                template: AppBundle:templates:community/Completion/completion-form.html.twig
+                type: AppBundle\Form\Type\CompletionType
+```
+
+## email
+
+Its possible that an email is also sent after the completion.
```
Comment (danrot) [inline Resources/doc/11-completion.md]: `It's` or `It is`

File: Resources/doc/11-completion.md [hunk 7]
```diff
@@ -0,0 +1,210 @@
...
+{% extends "AppBundle::master-email.html.twig" %}
+
+{% block content %}
+    You successfully completed your data.
+{% endblock %}
+```
+
+## service
+
+The service which validates the user data and check if a completion form should be displayed.
```
Comment (danrot) [inline Resources/doc/11-completion.md]: `and checks if a completion`

File: Resources/doc/12-blacklisting.md [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+                    user_template: AppBundle:templates:community/Registration/registration-email.html.twig
+                template: AppBundle:templates:community/Blacklist/blacklist-confirmed.html.twig
+```
+
+## Backend Config
+
+To enable the blacklisting feature add the permissions for blacklisting to your role.  
+For this go to `Settings -> Roles -> YourRole` in the sulu backend and add the permissions.  
+  
+When you successfully enabled it you can `denied` or set a email to `request` under `Settings -> Blacklist`.
```
Comment (danrot) [inline Resources/doc/12-blacklisting.md]: `you can denied` sound strange... Is `denied` a state?

File: Resources/doc/12-blacklisting.md [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+                template: AppBundle:templates:community/Blacklist/blacklist-confirmed.html.twig
+```
+
+## Backend Config
+
+To enable the blacklisting feature add the permissions for blacklisting to your role.  
+For this go to `Settings -> Roles -> YourRole` in the sulu backend and add the permissions.  
+  
+When you successfully enabled it you can `denied` or set a email to `request` under `Settings -> Blacklist`. 
+It is possible to use wildcard e.g.: `*@test.com` to set the type for a whole domain.
```
Comment (danrot) [inline Resources/doc/12-blacklisting.md]: `use wildcards`

Comment (danrot) [inline Resources/doc/12-blacklisting.md]: Remove the `:`

File: Resources/doc/12-blacklisting.md [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+    <p>Confirm: <a href="{{ confirmUrl }}">{{ confirmUrl }}</a></p>
+
+    {% set denyUrl = url('sulu_community.user_deny', { token: token }) %}
+    <p>Deny: <a href="{{ denyUrl }}">{{ denyUrl }}</a></p>
+{% endblock %}
+```
+
+## blacklist_denied.template / blacklist_confirmed.template
+
+When the admin click on the link a template is rendered which can show specific content:
```
Comment (danrot) [inline Resources/doc/12-blacklisting.md]: `clicks`

File: Resources/doc/12-blacklisting.md [hunk 11]
```diff
@@ -0,0 +1,96 @@
...
+{% extends "AppBundle::master.html.twig" %}
+
+{% block content %}
+    User "{{ user.email }}" denied/confirmed.
+{% endblock %}
+```
+
+## blacklist_confirmed.email
+
+If the user is confirmed he will receive a email with the confirmation link:
```
Comment (danrot) [inline Resources/doc/12-blacklisting.md]: `an email`

File: Resources/doc/2-setup-webspace.md [hunk 12]
```diff
@@ -0,0 +1,82 @@
...
+<!-- app/Resources/webspaces/<your_webspace>.xml -->
+
+<security>
+    <system>Website</system>
+</security>
+```
+
+## Activate Community Features
+
+Enable community features for webspace:
```
Comment (danrot) [inline Resources/doc/2-setup-webspace.md]: `for your webspace`

File: Resources/doc/3-customization.md [hunk 13]
```diff
@@ -0,0 +1,213 @@
...
+
+You can change the role name. 
+If not set it will generate it by webspace key and add 'User' as postfix.
+
+#### firewall
+
+Type: `string`
+Example: website
+
+Set the firewall for the current webspace else the firewall name need to be the same as the webspace key.
```
Comment (danrot) [inline Resources/doc/3-customization.md]: `needs to be`- And I don't really understand what that sentence means.

File: Resources/doc/3-customization.md [hunk 14]
```diff
@@ -0,0 +1,213 @@
...
+
+### Type Options
+
+#### activate_user
+
+Type: `bool`
+Example: true
+
+When set to true the user can login after successfully form submit.
```
Comment (danrot) [inline Resources/doc/3-customization.md]: `submit the form`

File: Resources/doc/3-customization.md [hunk 15]
```diff
@@ -0,0 +1,213 @@
...
+
+#### auto_login
+
+Type: `bool`
+Example: true
+
+The user will be automatically login after successfully form submit.
```
Comment (danrot) [inline Resources/doc/3-customization.md]: `automatically logged in after successfully submitting the form`

File: Resources/doc/3-customization.md [hunk 16]
```diff
@@ -0,0 +1,213 @@
...
+
+#### embed_template
+
+Type: `string`
+Example: AppBundle:templates:community/Login/login-embed.html.twig
+
+The template which when use `render_esi` to render a login.
```
Comment (danrot) [inline Resources/doc/3-customization.md]: `which when` doesn't make sense... `The template which is used by render_esi to render a login`?

File: Resources/doc/3-customization.md [hunk 17]
```diff
@@ -0,0 +1,213 @@
...
+
+#### email.subject
+
+Type: `string`
+Example: Registration
+
+The subject of the send emails.
```
Comment (danrot) [inline Resources/doc/3-customization.md]: `the sent emails` or even only `the emails` - it is obvious that they are sent.

File: Resources/doc/3-customization.md [hunk 18]
```diff
@@ -0,0 +1,213 @@
...
+
+#### email.admin_template
+
+Type: `string`
+Example: AppBundle:templates:community/Blacklist/blacklist-email.html.twig
+
+The template used to render the admin email if null no email is sent.
```
Comment (danrot) [inline Resources/doc/3-customization.md]: `if no email is sent`? `if null no email` doesn't sound right.

File: Resources/doc/3-customization.md [hunk 19]
```diff
@@ -0,0 +1,213 @@
...
+
+#### email.user_template
+
+Type: `string`
+Example: AppBundle:templates:community/Registration/registration-email.html.twig
+
+The template used to render the user email if null no email is sent.
```
Comment (danrot) [inline Resources/doc/3-customization.md]: Same as above

File: Resources/doc/3-customization.md [hunk 20]
```diff
@@ -0,0 +1,213 @@
...
+
+#### redirect_to
+
+Type: `string`
+Example: ?send=true, app.redirect_route
+
+Redirect to can be a url or a route_name. When contain `{localization}` it will be replaced with the current locale.
```
Comment (danrot) [inline Resources/doc/3-customization.md]: `redirect_to` can be a ... `When containing {localization}`

File: Resources/doc/3-customization.md [hunk 21]
```diff
@@ -0,0 +1,213 @@
...
+
+#### service
+
+Type: `string`
+Example: app.completion_validator
+
+Service from interface `CompletionInterface` to validate if a user need to add additional information before using your application.
```
Comment (danrot) [inline Resources/doc/3-customization.md]: `if a user needs to`

File: Resources/doc/4-login.md [hunk 22]
```diff
@@ -0,0 +1,103 @@
...
+    webspaces:
+        <webspace_key>:
+            login:
+                embed_template: AppBundle:templates:community/Login/login-embed.html.twig
+                template: AppBundle:templates:community/Login/login.html.twig
+```
+
+## embed_template
+
+The embed template can be used to display on every page a login or logout button based if a user is logged in or not.
```
Comment (danrot) [inline Resources/doc/4-login.md]: `based on the login state of the user`

File: Resources/doc/5-registration.md [hunk 23]
```diff
@@ -0,0 +1,179 @@
...
+                    subject: Registration
+                    user_template: AppBundle:templates:community/Registration/registration-email.html.twig
+                    admin_template: ~
+                template: AppBundle:templates:community/Registration/registration-form.html.twig
+                type: AppBundle\Form\Type\RegistrationType
+```
+
+## email
+
+The registration email mostly contains the confirmationKey.
```
Comment (danrot) [inline Resources/doc/5-registration.md]: `The registration email contains a confirmation link`?

File: Resources/doc/5-registration.md [hunk 24]
```diff
@@ -0,0 +1,179 @@
...
+
+    <a href="{{ url }}">
+        {{ url }}
+    </a>
+{% endblock %}
+```
+
+## template
+
+The template contains the form and mostly the success message.
```
Comment (danrot) [inline Resources/doc/5-registration.md]: omit `mostly`

File: Resources/doc/5-registration.md [hunk 25]
```diff
@@ -0,0 +1,179 @@
...
+        </p>
+    {% else %}
+        {{ form(form) }}
+    {% endif %}
+{% endblock %}
+```
+
+## type
+
+You can create your own form by set your own RegistrationType
```
Comment (danrot) [inline Resources/doc/5-registration.md]: `by setting` and end with a dot

File: Resources/doc/6-confirmation.md [hunk 26]
```diff
@@ -0,0 +1,67 @@
...
+
+    <a href="{{ url('sulu_community.login') }}">
+        Login
+    </a>
+{% endblock %}
+```
+
+## template
+
+When the user click on the confirmation link the following template will be render:
```
Comment (danrot) [inline Resources/doc/6-confirmation.md]: `clicks`

Comment (danrot) [inline Resources/doc/6-confirmation.md]: `rendered`

File: Resources/doc/7-password-forget.md [hunk 27]
```diff
@@ -0,0 +1,114 @@
...
+        </p>
+    {% else %}
+        {{ form(form) }}
+    {% endif %}
+{% endblock %}
+```
+
+## type
+
+Set a new type to overwrite the exist form.
```
Comment (danrot) [inline Resources/doc/7-password-forget.md]: `existing form`

File: Resources/doc/8-password-reset.md [hunk 28]
```diff
@@ -0,0 +1,118 @@
...
+                    admin_template: ~
+                    user_template: AppBundle:templates:community/Password/reset-email.html.twig
+                redirect_to: ?send=true
+                template: AppBundle:templates:community/Password/reset-form.html.twig
+                type: AppBundle\Form\Type\PasswordResetType
+```
+
+## email
+
+After the user submitted the password reset form he will receive a email.
```
Comment (danrot) [inline Resources/doc/8-password-reset.md]: `an email`

File: Resources/doc/8-password-reset.md [hunk 29]
```diff
@@ -0,0 +1,118 @@
...
+        {% else %}
+            The received token is invalid.
+        {% endif %}
+    {% endif %}
+{% endblock %}
+```
+
+## type
+
+Set a new type to overwrite the exist form.
```
Comment (danrot) [inline Resources/doc/8-password-reset.md]: `existing form`

File: Resources/doc/9-profile.md [hunk 30]
```diff
@@ -0,0 +1,344 @@
...
+                    user_template: ~
+                    admin_template: ~
+                redirect_to: ~
+                template: AppBundle:templates:community/Profile/profile-form.html.twig
+                type: AppBundle\Form\Type\ProfileType
+```
+
+## email
+
+Its possible that an email is also sent when a profile is changed.
```
Comment (danrot) [inline Resources/doc/9-profile.md]: `It's` or `It is`

File: Resources/doc/9-profile.md [hunk 31]
```diff
@@ -0,0 +1,344 @@
...
+        </p>
+    {% endif %}
+
+    {{ form(form) }}
+{% endblock %}
+```
+
+## type
+
+Set a new type to overwrite the exist form.
```
Comment (danrot) [inline Resources/doc/9-profile.md]: `existing form`

**Comment by danrot** [comment]
`Password forget` sound strange... `forget` is a verb, and should be before the noun and in the past: `Forgot password` (you can also check google search for the term).

I would like `Password reset` better, but it's already used somehow, what's the difference between `Password forget` and `Password reset`

**Comment by alexander-schranz** [comment]
@danrot 

Password Forget is the process where you input your username/email and normally get an email with a link (it could also be configured that only an administrator will get an email).

Password Reset is the process where you set the new password.

File: Resources/doc/2-setup-webspace.md [hunk 32]
```diff
@@ -0,0 +1,82 @@
...
+# app/config/website/security.yml
+
+security:
+    session_fixation_strategy: none
+
+    access_decision_manager:
+        strategy: affirmative
+
+    acl:
+        connection: default
```
Comment (alexander-schranz) [inline Resources/doc/2-setup-webspace.md]: was removed in https://github.com/sulu/sulu-minimal/pull/34 check if this is still needed.

File: Resources/doc/1-installation.md [hunk 33]
```diff
@@ -0,0 +1,50 @@
...
+public function registerBundles()
+{
+    $bundles = [
+        // ...
+        new Symfony\Bundle\SecurityBundle\SecurityBundle(),
+        new Sulu\Bundle\CommunityBundle\SuluCommunityBundle(),
+    ];
+}
+```
+
```
Comment (alexander-schranz) [inline Resources/doc/1-installation.md]: Remove the SecurityBundle from AdminKernel to avoid error.

File: README.md (hunk 3 above)
Comment (alexander-schranz) [inline README.md]: @danrot 

The whole Password Forget/Reset has 2 processes:

**Password Forget process**

If you click on "I forgot my password" enter your email/username and an email will be sent (to you or to the adminstrator). 

**Password Reset process**

You or the the admin click a link. Set the new password and an email will sent to you or to an administrator that the new password was set.

Comment (alexander-schranz) [inline README.md]: will call it in the feature list `Password Forget/Reset`

File: Resources/doc/12-blacklisting.md (hunk 8 above)
Comment (alexander-schranz) [inline Resources/doc/12-blacklisting.md]: @danrot it is `block` and `request` which are the state what you think about:

When you successfully enabled it email addresses can be set to `block` or on `request` under `Settings -> Blacklist`.

File: Resources/doc/3-customization.md (hunk 13 above)
Comment (alexander-schranz) [inline Resources/doc/3-customization.md]: @danrot its a fallback mechanism when no `firewall` is configured the webspaceKey is used.

File: Resources/doc/3-customization.md (hunk 17 above)
Comment (alexander-schranz) [inline Resources/doc/3-customization.md]: @danrot what about:

`The template used to render the admin email, set to null to deactivate it.`

**Comment by alexander-schranz** [comment]
@wachterjohannes @danrot documentation updated.

**Pull request merged**